File: MissionControl/runtime/python/memory/embeddings.py
"""
Embeddings provider for Python agents
Supports OpenAI and Groq
"""

import os
from abc import ABC, abstractmethod
from typing import List
import numpy as np
from openai import OpenAI, AsyncOpenAI
from groq import Groq, AsyncGroq
import hashlib

class EmbeddingProvider(ABC):
    @abstractmethod
    def embed_text(self, text: str) -> List[float]:
        pass

    @abstractmethod
    def embed_batch(self, texts: List[str]) -> List[List[float]]:
        pass

class OpenAIEmbeddings(EmbeddingProvider):
    def __init__(self, api_key: str = None, model: str = "text-embedding-ada-002"):
        self.api_key = api_key or os.getenv("OPENAI_API_KEY")
        if not self.api_key:
            raise ValueError("OPENAI_API_KEY is required")
        self.model = model
        self.client = OpenAI(api_key=self.api_key)
        self._cache = {}

    def _cache_key(self, text: str):
        h = hashlib.sha256(text.encode()).hexdigest()[:16]
        return f"{self.model}:{h}"

    def embed_text(self, text: str) -> List[float]:
        text = text[:8000]  # limit
        cache_key = self._cache_key(text)
        if cache_key in self._cache:
            return self._cache[cache_key]

        response = self.client.embeddings.create(input=text, model=self.model)
        embedding = response.data[0].embedding
        self._cache[cache_key] = embedding
        return embedding

    def embed_batch(self, texts: List[str]) -> List[List[float]]:
        results = []
        for text in texts:
            results.append(self.embed_text(text))
        return results

class GroqEmbeddings(EmbeddingProvider):
    def __init__(self, api_key: str = None, model: str = "text-embedding-ada-002"):
        self.api_key = api_key or os.getenv("GROQ_API_KEY")
        if not self.api_key:
            raise ValueError("GROQ_API_KEY is required")
        self.model = model
        self.client = Groq(api_key=self.api_key)
        self._cache = {}

    def _cache_key(self, text: str):
        h = hashlib.sha256(text.encode()).hexdigest()[:16]
        return f"{self.model}:{h}"

    def embed_text(self, text: str) -> List[float]:
        text = text[:8000]
        cache_key = self._cache_key(text)
        if cache_key in self._cache:
            return self._cache[cache_key]

        response = self.client.embeddings.create(input=text, model=self.model)
        embedding = response.data[0].embedding
        self._cache[cache_key] = embedding
        return embedding

    def embed_batch(self, texts: List[str]) -> List[List[float]]:
        results = []
        for text in texts:
            results.append(self.embed_text(text))
        return results

def get_embedding_provider() -> EmbeddingProvider:
    provider = os.getenv("EMBEDDING_PROVIDER", "openai").lower()
    if provider == "openai":
        return OpenAIEmbeddings()
    elif provider == "groq":
        return GroqEmbeddings()
    else:
        raise ValueError(f"Unknown embedding provider: {provider}")

File: MissionControl/src/memory/embeddings.js
/**
 * Embeddings Service for Virtual Agency Memory
 * Supports OpenAI and Groq embeddings models
 */

import dotenv from 'dotenv';
dotenv.config();

const DEFAULT_MODEL = 'text-embedding-ada-002';
const DIMENSION = 1536;

// Cache para evitar llamadas repetitivas
const embeddingCache = new Map();

/**
 * Generate embedding for a single text
 * @param {string} text - Text to embed
 * @param {object} options - { provider: 'openai'|'groq', apiKey?: string, model?: string }
 * @returns {Promise<number[]>}
 */
export async function generateEmbedding(text, options = {}) {
  const {
    provider = process.env.EMBEDDING_PROVIDER || 'openai',
    apiKey = process.env[`${provider.toUpperCase()}_API_KEY`],
    model = process.env.EMBEDDING_MODEL || DEFAULT_MODEL
  } = options;

  if (!apiKey) {
    throw new Error(`Missing API key for ${provider}. Set ${provider.toUpperCase()}_API_KEY env var.`);
  }

  // Normalize and cache key
  const normalized = text.trim().slice(0, 8000); // limit input
  const cacheKey = `${provider}:${model}:${normalized.slice(0, 200)}`;

  if (embeddingCache.has(cacheKey)) {
    return embeddingCache.get(cacheKey);
  }

  let embedding;
  if (provider === 'openai') {
    embedding = await callOpenAI(text, apiKey, model);
  } else if (provider === 'groq') {
    embedding = await callGroq(text, apiKey, model);
  } else {
    throw new Error(`Unknown embedding provider: ${provider}`);
  }

  embeddingCache.set(cacheKey, embedding);
  return embedding;
}

/**
 * Generate embeddings for multiple texts (batch)
 * @param {string[]} texts
 * @param {object} options
 * @returns {Promise<number[][]>}
 */
export async function generateEmbeddingsBatch(texts, options = {}) {
  const batchSize = parseInt(process.env.EMBEDDING_BATCH_SIZE || '100');
  const results = [];

  for (let i = 0; i < texts.length; i += batchSize) {
    const batch = texts.slice(i, i + batchSize);
    const embeddings = await Promise.all(batch.map(text => generateEmbedding(text, options)));
    results.push(...embeddings);

    // Small delay to avoid rate limits
    if (i + batchSize < texts.length) {
      await new Promise(resolve => setTimeout(resolve, 100));
    }
  }

  return results;
}

async function callOpenAI(text, apiKey, model) {
  const response = await fetch('https://api.openai.com/v1/embeddings', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${apiKey}`
    },
    body: JSON.stringify({
      input: text,
      model: model
    })
  });

  if (!response.ok) {
    const err = await response.text();
    throw new Error(`OpenAI embedding error: ${response.status} - ${err}`);
  }

  const data = await response.json();
  return data.data[0].embedding;
}

async function callGroq(text, apiKey, model) {
  // Groq uses OpenAI-compatible endpoint for embeddings
  const response = await fetch('https://api.groq.com/openai/v1/embeddings', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${apiKey}`
    },
    body: JSON.stringify({
      input: text,
      model: model || 'text-embedding-ada-002' // Groq supports this model via OpenAI API
    })
  });

  if (!response.ok) {
    const err = await response.text();
    throw new Error(`Groq embedding error: ${response.status} - ${err}`);
  }

  const data = await response.json();
  return data.data[0].embedding;
}

/**
 * Clear embedding cache (useful for testing or forced refresh)
 */
export function clearEmbeddingCache() {
  embeddingCache.clear();
}

/**
 * Get cache statistics
 */
export function getEmbeddingCacheStats() {
  return {
    size: embeddingCache.size,
    keys: Array.from(embeddingCache.keys()).slice(0, 10) // sample
  };
}

